x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/runtimefields/RuntimeFieldsFeatureSetUsage.java

```diff
@@ -0,0 +1,272 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RuntimeFieldsFeatureSetUsage extends XPackFeatureSet.Usage {
+
+    public static RuntimeFieldsFeatureSetUsage fromMetadata(Iterable<IndexMetadata> metadata) {
+        Map<String, RuntimeFieldStats> fieldTypes = new HashMap<>();
+        for (IndexMetadata indexMetadata : metadata) {
+            if (indexMetadata.isSystem()) {
+                // Don't include system indices in statistics about mappings, we care about the user's indices.
+                continue;
+            }
+            Set<String> indexFieldTypes = new HashSet<>();
+            MappingMetadata mappingMetadata = indexMetadata.mapping();
+            if (mappingMetadata != null) {
+                Object runtimeObject = mappingMetadata.getSourceAsMap().get("runtime");
+                if (runtimeObject instanceof Map) {
+                    @SuppressWarnings("unchecked")
+                    Map<String, Object> runtimeMappings = (Map<String, Object>) runtimeObject;
+                    for (Object runtimeFieldMappingObject : runtimeMappings.values()) {
+                        if (runtimeFieldMappingObject instanceof Map) {
+                            @SuppressWarnings("unchecked")
+                            Map<String, Object> runtimeFieldMapping = (Map<String, Object>) runtimeFieldMappingObject;
+                            Object typeObject = runtimeFieldMapping.get("type");
+                            if (typeObject != null) {
+                                String type = typeObject.toString();
+                                RuntimeFieldStats stats = fieldTypes.computeIfAbsent(type, RuntimeFieldStats::new);
+                                stats.count++;
+                                if (indexFieldTypes.add(type)) {
+                                    stats.indexCount++;
+                                }
+                                Object scriptObject = runtimeFieldMapping.get("script");
+                                if (scriptObject == null) {
+                                    stats.scriptLessCount++;
+                                } else if (scriptObject instanceof Map) {
+                                    @SuppressWarnings("unchecked")
+                                    Map<String, Object> script = (Map<String, Object>) scriptObject;
+                                    Object sourceObject = script.get("source");
+                                    if (sourceObject != null) {
+                                        String scriptSource = sourceObject.toString();
+                                        int chars = scriptSource.length();
+                                        long lines = scriptSource.lines().count();
+                                        int docUsages = countOccurrences(scriptSource, "doc[\\[\\.]");
```

Comment:
I think its worth double checking if this matches stuff like `doc\`. I never can remember the escaping rules for character classes.

Reply:
I think it's ok, but I am not sure I can be trusted, I will double check